### ISSUE TITLE
[pdata] Find replacement for now deprecated gogo/protobuf

### ISSUE THREAD

**Issue by yurishkuro** [open-telemetry/opentelemetry-proto#446]
I don't know the full background of using [gogo/protobuf](https://github.com/gogo/protobuf), I suspect it was similar to Jaeger's use of gogoproto for serialization efficiency (in particular, the ability to embed structs by value, as in `[]KeyValue` instead of `[]*KeyValue`, which significantly reduces memory allocations).

Well, the bad news, which was long coming, is that GoGo is officially deprecated since Oct'22. Continue to depend on it going forward presents security and compatibility risks.

This ticket is to kick off a discussion for a viable path forward, since I don't have a solution in mind right now.

Some past related tickets:
  * https://github.com/open-telemetry/opentelemetry-collector/issues/4258

**Comment by tigrannajaryan** [comment]
Yes, Collector uses Gogoproto instead of Google's implementation for performance reasons.

Here are a few replacement options that I can think of:

- Move to Google Protobuf library. This will most likely result in worse performance for the the Collector (about 1.3x worse in our past benchmarks IIRC). The newer version of Google Protobuf (v2?) is even slower than the older version.
- Otel to take responsibility of entire Gogoproto. This is a big undertaking, I don't think we want to subscribe to that.
- Fork Gogoproto and maintain an Otel fork just for ourselves. It is still a pretty large codebase to maintain and my cursory assessment of it was that it was not a codebase that I would want to take responsibility to maintain myself.
- Create our own Protobuf library. This may sound crazy but may be tractable if we limit it to only support what OTLP needs. I managed to write an [experimental compiler/generator](https://github.com/splunk/exp-lazyproto) in just a few days that works on OTLP Logs and may be faster than Gogoproto (and with optional fields support, BTW). Obviously a production quality library requires much bigger investment and I definitely won't claim that I alone want to be a sole author and maintainer of such a library.

Unfortunately I do not see great options here, all have some downsides. We will need to decide how much effort we are willing to invest into this.
